README.md
SolidRuby
===========

SolidRuby is a framework for programming 2d and 3d OpenSCAD models in Ruby.
It is a fork of [CrystalScad](https://github.com/Joaz/CrystalScad) with the aim to update and improve the codebase.

SolidRuby, like CrystalScad and RubyScad on which it is based, is released under the GPLv3 license.

[![Build Status](https://travis-ci.org/MC-Squared/SolidRuby.svg?branch=master)](https://travis-ci.org/MC-Squared/SolidRuby)

Installation:
===========

Dependencies:

- Ruby 1.9.3+

Install via gem:
```
# gem install solidruby
```

if you have multiple ruby versions, you likely need to use gem1.9.3 instead of gem.

Install via git:

- clone repository
- \# rake install

Getting started
===========
SolidRuby comes with a generator that generates project stubs automatically for you. Run this command from a terminal in the directory that you want to create a project:

```
# solidruby new [my_project_name]
```
Change [my_project_name] to the name of your project

A project named "my_project" will create those files and directories:

- my_project/my_project.rb - the controller
- my_project/lib/assemblies - for putting together assemblies of individual parts
- my_project/lib/electronics - put electronics here
- my_project/lib/hardware - put hardware parts in here
- my_project/lib/printed - put parts that you want to print in here
- my_project/lib/lasercut - put sheets that need to be cut (by laser or other) in here
- my_project/lib/params.rb - place for global parameters
- my_project/lib/assemblies/my_project_assembly.rb  - dummy assembly
- my_project/Guardfile - guard

A Guardfile is created to automatically update the output scad files.

Coding
===========
Nearly all OpenSCAD functions are implemented. You can use the same parameters as in OpenSCAD, although SolidRuby provides some convenient variable names like diameter (d) for cylinders.

Some examples:

CSG Modeling:
```
res = cylinder(d:10, h:10)
# union
res += cube(x:5, y:20, z:20)
# difference
res -= cylinder(d:5, h:10)
# intersection
res *= cylinder(d:10, h:10)
```

Chain transformations:  
```
res = cube(1, 2, 3).rotate(x: 90).translate(x: 20, y: 2, z: 1).mirror(z: 1)
```


Hull:   
```
  res = hull(cylinder(d:10, h:10).cube(20, 10, 10].translate(x: 10)))
```

Center cubes in X/Y direction only:    
```
  cube(10, 10, 10).center_xy # note: only works on cubes and must be put before any transformations
```

Also implemented: center_x, center_y, center_z, center

Helpers:

Some helpers have been implemented to ease modelling.
Where helpers take face/edge/corner the options are:
```
  faces: :top, :bottom, :left, :right, :front, :back, :center
  edges: :top, :bottom, :left, :right, :front, :back, :center
  corners: :left, :right, :center
```
For example:

Triangles - can be either linear_extruded or simply used for calulations
  triangle(alpha: 90, c: 5, b: 3) #a right triangle on the origin, with Y leg 5 units, X leg 3 units
```
triangle(alpha: 45, c: 6, b: 9).beta #get calulated angle from triangle
```

Filleted cubes
Filled cubes can be created by using the .fillet helper:
```
cube(10).fillet(edges: :vertical, r: 2)
cube(10).fillet(top: [:left, :right], front: :bottom, r: 2)
```

Chamfered cubes
Similar to filleted cubes above
```
cube(10).chamfer(edges: :vertical, h: 2)
```

Both the fillet() and chamfer() helper can be used in isolation as well:
```
fillet(h: 10, r: 2)
chamfer(l: 10, h: 2)
```

Rounded cubes:
Due to how fillet is applied, the corners of a cube are not perfect when doing all sides.
In this case you can use rounded_cube:
```
rounded_cube(x: 10, y: 10, z: 10, r: 2) #Actually a hull of 8 spheres
```

Hollow cylinders:
Hollow cylinders can be created by indicating an :id argument (inner diameter)
```
cylinder(d: 10, id: 8, h: 10) #creates a fully hollow cylinder
cylinder(d: 10, id: 8, h: 10, ih: 8) #inner cutout only partial height
```

Placing onto:
For simple primitives you can use .place to place objects onto their faces:
```
  c = cube(10)
  res = c + cube(5).place(onto: c, face: :top, edge: :top, corner: :center)

  cy = cylinder(d: 10, h: 50)
  res = cy + sphere(10).place(onto: cy, face: :top) #edge and corner default to center
```

Long slots:   
```
  # produces a hull of two cylinders, 14mm apart        
  long_slot(d:4.4, h:10, l:14)  
```

Printed threads:
There is a printed thread class for printing threads, i.e. for nuts and bolts:
```
t1 = PrintedThread.new(diameter: 20, pitch: 2.5, length: 22.5, internal: true)
```


A few tips:
- Be visual. By running `guard` you can watch your updates in OpenSCAD

- When porting OpenSCAD code, beware of dividing integers. Example:
```
cylinder(r=11/2, h=10);
```
needs to be ported to
```
cylinder(r:11/2.0, h:10)
```
or
```
cylinder(d:11, h:10)
```


License:
===========
GPLv3


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.3.0] - 2022-10-01

### Added
- Cube now accepts `center: [:x, :y, :z]` arguments
- Add `.mirror(:x, :y, :z)` method

### Changed

- `SolidRubyObject#translate` now combines consecutive translations
- Chamfers and Fillets can now be chained e.g. `cube.fillet(...).fillet(...)`

### Removed
- Remove deprecated `stack` method

## [0.2.0] - 2020-06-12
### Added
- New way of handling parameters via `params` singleton + `parameters.yml` file
- New `ProjectCli` class to allow overriding parameters from command line
- Ruby exceptions are now output in OpenSCAD view

### Changed

- Minimum Ruby version bumped to 3.0
- Runtime dependency of `wijet-thor` replaced with `thor`
- Upgrade to bundler v2

### Fixed
- Fix Import and LinearExtrude incorrectly accepting scale arguments

examples/threads.rb
#!/usr/bin/env ruby

require 'solidruby'
include SolidRuby

# This example shows the use of ScrewThreads in SolidRuby.

class BlackBox < SolidRuby::Assembly
  # This is a box (the part that you want to make a mount for)

  def initialize
    @x = 200
    @y = 100
    @z = 50
  end

  #	note: when you design a box like that, make sure that it starts at 0,0,0
  def show
    cube([@x, @y, @z]).color(r: 10, b: 10, g: 10, a: 150)
  end

  # We're defining threads on every side. It must be called thread_side
  # where side is either top, bottom, left, right, front, back
  def threads_top
    holes = *ScrewThread.new(x: 50, y: 50, size: 8, depth: 10) # note that =* creates an Array
    holes << ScrewThread.new(x: 150, y: 50, size: 8, depth: 10)	# and << adds stuff to an Array
  end

  def threads_bottom
    [ScrewThread.new(x: 50, y: 50, size: 8, depth: 10)]
  end

  # note that the coordinates for the threads are needed in 2 directions from 0,0,0
  # on top and bottom you're defining x & y
  # on left and right you're defining y & z
  # on front and back you're defining x & z
  def threads_left
    holes = *ScrewThread.new(y: 15, z: 10, size: 3, depth: 5)
    holes << ScrewThread.new(y: 40, z: 30, size: 3, depth: 5)
  end

  def threads_right
    holes = *ScrewThread.new(y: 15, z: 10, size: 3, depth: 5)
    holes << ScrewThread.new(y: 40, z: 30, size: 3, depth: 5)
  end

  # this example has different thread sizes for m3 and m6
  def threads_front
    holes = *ScrewThread.new(x: 20, z: 10, size: 3, depth: 5)
    holes << ScrewThread.new(x: 100, z: 20, size: 6, depth: 25)
    holes << ScrewThread.new(x: 140, z: 20, size: 6, depth: 25)
    holes << ScrewThread.new(x: 180, z: 45, size: 3, depth: 5)
  end

  def threads_back
    holes = *ScrewThread.new(x: 20, z: 10, size: 3, depth: 5)
    holes << ScrewThread.new(x: 100, z: 25, size: 12, depth: 25)
    holes << ScrewThread.new(x: 180, z: 45, size: 3, depth: 5)
  end
end

# this is what a basic mount looks like.
class BlackBoxMountTop < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: 200, y: 100, z: 7)
    # the mount should either be a cube or an object that returns x,y,z as dimensions
    # if you have a complicated part, split the cube and add it later on to your object
    # Alternatively, define height in the arguments.

    # create_bolts needs the face as first argument, your new mount as second argument
    # and the object that has the threads as third argument
    # it accepts an hash of arguments as 4th argument. You can define
    #	height: define a custom height of your mount
    # bolt_height: custom height for your bolt(s). Can also be an array for different lengths.
    #							 This is useful if the automatic length calculation doesn't produce values
    #							 that are available.
    # 						 This example would, without the setting 16 produce a M8x17 bolt. We don't have
    #							 that size, so we use M8x16 instead.

    bolts = create_bolts('top', mount, box, bolt_height: 16)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(z: box.z)
    res += box.show if show
    res
  end
end

class BlackBoxMountBottom < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: 200, y: 100, z: z = 2)
    # note that the bolt height bom output is unchecked in this and the following examples
    bolts = create_bolts('bottom', mount, box)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(z: -z)
    res += box.show if show
    res
  end
end

class BlackBoxMountLeft < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: x = 5, y: 100, z: 50)

    bolts = create_bolts('left', mount, box)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(x: -x)
    res += box.show if show
    res
  end
end

class BlackBoxMountRight < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: 5, y: 100, z: 50)

    bolts = create_bolts('right', mount, box)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(x: box.x)
    res += box.show if show
    res
  end
end

class BlackBoxMountFront < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: box.x, y: y = 15, z: box.z)

    bolts = create_bolts('front', mount, box)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(y: -y)
    res += box.show if show
    res
  end
end

class BlackBoxMountBack < SolidRuby::Printed
  def part(show)
    box = BlackBox.new
    mount = cube(x: box.x, y: 35, z: box.z)

    bolts = create_bolts('back', mount, box)
    mount -= bolts
    mount += bolts if show

    res = mount.translate(y: box.y)
    res += box.show if show
    res
  end
end

b = SolidRubyObject.new

# uncomment as many mounts as you like here

b += BlackBoxMountTop.new.show
b +=BlackBoxMountBottom.new.show
b +=BlackBoxMountLeft.new.show
b +=BlackBoxMountRight.new.show
b +=BlackBoxMountFront.new.show
b +=BlackBoxMountBack.new.show

# you can also try output instead of show
# b =BlackBoxMountTop.new.output

# Uncomment this for checking the BOM output
# puts BillOfMaterial.bom.output

b.save('threads.scad')
